velox/experimental/cudf/exec/ExpressionEvaluator.cpp

```diff
@@ -820,4 +833,365 @@ bool ExpressionEvaluator::canBeEvaluated(
+    if (range->isSingleValue()) {
+      // Only one value in the IN-list, so just compare for equality.
+      auto singleValue = range->lower();
+      const auto& literal = addLiteral(singleValue);
+      return tree.push(Operation{Op::EQUAL, columnRef, literal});
+    }
+  }
+
+  if (!lowerUnbounded) {
+    auto lowerValue = range->lower();
+    const auto& lowerLiteral = addLiteral(lowerValue);
+
+    auto lowerOp = range->lowerExclusive() ? Op::GREATER : Op::GREATER_EQUAL;
+    lowerExpr = &tree.push(Operation{lowerOp, columnRef, lowerLiteral});
+  }
+
+  if (!upperUnbounded) {
+    auto upperValue = range->upper();
+    const auto& upperLiteral = addLiteral(upperValue);
+
+    auto upperOp = range->upperExclusive() ? Op::LESS : Op::LESS_EQUAL;
+    upperExpr = &tree.push(Operation{upperOp, columnRef, upperLiteral});
+  }
+
+  if (lowerExpr && upperExpr) {
+    return tree.push(Operation{Op::NULL_LOGICAL_AND, *lowerExpr, *upperExpr});
+  } else if (lowerExpr) {
+    return *lowerExpr;
+  } else if (upperExpr) {
+    return *upperExpr;
+  }
+
+  // Both bounds unbounded => Pass-through filter (everything).
+  return tree.push(Operation{Op::EQUAL, columnRef, columnRef});
+}
+
+template <TypeKind Kind>
+std::reference_wrapper<const cudf::ast::expression> buildBigintRangeExpr(
+    const common::Filter& filter,
+    cudf::ast::tree& tree,
+    std::vector<std::unique_ptr<cudf::scalar>>& scalars,
+    const cudf::ast::expression& columnRef,
+    rmm::cuda_stream_view stream,
+    rmm::device_async_resource_ref mr,
+    const TypePtr& columnTypePtr) {
+  using NativeT = typename TypeTraits<Kind>::NativeType;
+
+  if constexpr (std::is_integral_v<NativeT>) {
+    using Op = cudf::ast::ast_operator;
+    using Operation = cudf::ast::operation;
+
+    auto* bigintRange = static_cast<const common::BigintRange*>(&filter);
+
+    const auto lower = bigintRange->lower();
+    const auto upper = bigintRange->upper();
+
+    const bool skipLowerBound =
+        lower <= static_cast<int64_t>(std::numeric_limits<NativeT>::min());
+    const bool skipUpperBound =
+        upper >= static_cast<int64_t>(std::numeric_limits<NativeT>::max());
+
+    auto addLiteral = [&](int64_t value) -> const cudf::ast::expression& {
+      variant veloxVariant = static_cast<NativeT>(value);
+      const auto& literal =
+          makeScalarAndLiteral<Kind>(columnTypePtr, veloxVariant, scalars);
+      return tree.push(literal);
+    };
+
+    if (bigintRange->isSingleValue()) {
+      // Equal comparison: column = value. This value is the same as the
+      // lower/upper bound.
+      if (skipLowerBound || skipUpperBound) {
+        // If the singular value of this filter lies outside the range of the
+        // column's NativeT type then we want to be always false
+        return tree.push(Operation{Op::NOT_EQUAL, columnRef, columnRef});
+      } else {
+        auto const& literal = addLiteral(lower);
+        return tree.push(Operation{Op::EQUAL, columnRef, literal});
```

Comment:
Since the `columnRef` can be any integral type, we need to add `cast` operator if it is not int64 type, to compare with `int64` literal. Otherise AST will throw
```
E20250821 02:08:47.042325  9895 Exceptions.h:66] Line: /velox/velox/exec/Driver.cpp:574, Function:operator(), Expression:  Operator::getOutput failed for [operator: TableScan, plan node ID: 0]: CUDF failure at:/velox/_build/release/_deps/cudf-src/cpp/src/ast/expression_parser.cpp:244: An AST expression was provided non-matching operand types., Source: RUNTIME, ErrorCode: INVALID_STATE
```

Reply:
This breaks Q16 in velox cudf tpch benchmark.

Reply:
The literal for AST comparison operator is int64.  For bigInt range, the column used to compare may not be int64. it could be any other integral. AST will fail if the operands compared are not of same type. So, the column needs to be casted to int64.

Reply:
This is not supposed to happen as the `addLiteral` function only makes a literal of the same type as the column. See that the literal is of NativeT type which is the type of the column when dispatched. Moreover, this scenario is tested in unit tests https://github.com/devavret/velox/blob/33adf15ba797388f105befa99750a3f36a90d7d2/velox/experimental/cudf/tests/SubfieldFilterAstTest.cpp#L358-L378
Besides, this PR does not wire in this functionality in the table scans. Are you sure this conversion function was used?

Reply:
In this PR, it's not used in tablescan either.